File: src/main/java/net/ccbluex/liquidbounce/utils/ClientUtils.java
/*
 * LiquidCat Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CatsDevelopment/LiquidCat
 */
package net.ccbluex.liquidbounce.utils;

import net.minecraft.client.settings.GameSettings;
import net.minecraft.network.NetworkManager;
import net.minecraft.network.login.client.C01PacketEncryptionResponse;
import net.minecraft.network.login.server.S01PacketEncryptionRequest;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import javax.crypto.SecretKey;
import java.lang.reflect.Field;
import java.security.PublicKey;

@SideOnly(Side.CLIENT)
public final class ClientUtils extends MinecraftInstance {
    private static Field fastRenderField;

    static {
        try {
            fastRenderField = GameSettings.class.getDeclaredField("ofFastRender");

            if(!fastRenderField.isAccessible())
                fastRenderField.setAccessible(true);
        }catch(final NoSuchFieldException ignored) {
        }
    }

    public static void disableFastRender() {
        try {
            if(fastRenderField != null) {
                if(!fastRenderField.isAccessible())
                    fastRenderField.setAccessible(true);

                fastRenderField.setBoolean(mc.gameSettings, false);
            }
        }catch(final IllegalAccessException ignored) {
        }
    }

    public static void sendEncryption(final NetworkManager networkManager, final SecretKey secretKey, final PublicKey publicKey, final S01PacketEncryptionRequest encryptionRequest) {
        networkManager.sendPacket(new C01PacketEncryptionResponse(secretKey, publicKey, encryptionRequest.getVerifyToken()), p_operationComplete_1_ -> networkManager.enableEncryption(secretKey));
    }
}

File: src/main/kotlin/lol/liquidcat/utils/ClientUtils.kt
/*
 * LiquidCat Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CatsDevelopment/LiquidCat
 */
package lol.liquidcat.utils

import net.minecraft.client.Minecraft
import net.minecraft.client.settings.GameSettings
import net.minecraft.network.Packet
import net.minecraft.util.ChatComponentText
import java.lang.reflect.Field

object ClientUtils {
    private var fastRenderField: Field? = null

    init {
        runCatching {
            fastRenderField = GameSettings::class.java.getDeclaredField("ofFastRender")

            if (!fastRenderField!!.isAccessible)
                fastRenderField!!.isAccessible = true
        }
    }

    /**
     * Disables OptiFine FastRender
     */
    @JvmStatic
    fun disableFastRender() {
        runCatching {
            if (fastRenderField != null) {
                if (!fastRenderField!!.isAccessible)
                    fastRenderField!!.isAccessible = true

                fastRenderField!!.setBoolean(mc.gameSettings, false)
            }
        }
    }
}

/**
 * Minecraft instance
 */
val mc = Minecraft.getMinecraft()!!

/**
 * Sends client-side [message] to chat
 */
fun msg(message: String) {
    mc.ingameGUI.chatGUI.printChatMessage(ChatComponentText(message))
}

/**
 * Sends a [packet] to the server
 */
fun sendPacket(packet: Packet<*>) = mc.netHandler?.addToSendQueue(packet)